### DISCUSSION TITLE
Compression and Normalization problem & FLAC stream

### DISCUSSION THREAD

**Discussion by K1oma** [savonet/liquidsoap#3926] [User Support]
1. I've been struggling with normalization and dynamics for my radio station. I've tried numerous ways to normalize and compress and I've tried all the values possible to set the parameters. I've got no success in it, it doesn't sound like the result I want to hear. I would be grateful if somebody could show me a config of their compression and dynamics because I just loop and can't make progress, but I don't want to give up on an idea for my personal radio station.

2. FLAC format is the other thing I wanted to stream on my other Liquidsoap radio station. I used to ask ChatGPT to help me with the config, but even it doesn't give me a working config. I am not much of a programmer, but i've followed docs and PDF book, and that still doesn't help because I'm on a version **Liquidsoap 2.2.5+dev**

If someone could help me reach my goal, I would be grateful!

**Comment by vitoyucepi** [comment]
Hi @K1oma,

1. I don't know much about normalization and compression, but I can try to summon @Moonbase59.
   At least I know that you can analyze the volume using https://github.com/Moonbase59/loudgain and write tags that can be used with [the `replaygain` function](https://www.liquidsoap.info/doc-dev/replay_gain.html).
   There's also a very recent autocue feature, but the documentation is very limited. Here's a repository for reference https://github.com/Moonbase59/autocue.
1. > I used to ask ChatGPT to help me with the config, but even it doesn't give me a working config.

   This is just a chatbot, not a programming tool. It can generate some sequences of symbols that resemble the program, but the code doesn't necessarily do the expected thing.
   You can reliably get good results for popular things that are widely used, like natural language. On the other hand, you can reliably get bad results for niche things, like the liquidsoap programming language or ffmpeg arguments.

   > I used to ask ChatGPT to help me with the config, but even it doesn't give me a working config. I am not much of a programmer, but i've followed docs and PDF book, and that still doesn't help

   What's your current config? It's hard to predict what kind of AI-generated hallucinations you have.

**Comment by Moonbase59** [comment]
@K1oma That’s a topic I could talk hours and hours about… ;-) But it much depends on what you want to achieve, and what kind of music you play (Is it Ambient? Rock/Pop? Electronic Dance Music?). Do you prefer quality and good dynamic range over being "the loudest station on the dial"? Etc., etc…

Basically, I’d recommend building in steps:

1. Pre-process your tracks, adding _tags_ with _loudness information_, which will not in any way modify your valuable audio but allow the playout chain to adjust for same loudness playout. _Don ’t_ use obsolete old tools like _MP3Gain_, but something like [_loudgain_](https://github.com/Moonbase59/loudgain) and/or _rgbpm_, its mass-tagging tool. You only have to do this once. The typical loudness reference is -18 LUFS.

2. Use Liquidsoap’s `amplify` to adjust the playout loudness, ideally before any crossfading:
   ```
   radio = amplify(1., override="replaygain_track_gain", radio)
   ```
   This will assure all tracks being played at the same loudness, using the _ReplayGain Track Gain_ value stored in the audio files.

3. If you _cannot_ pre-process your tracks, for whatever reason, you can also use [`autocue.cue_file`](https://github.com/Moonbase59/autocue/tree/integrate-with-liquidsoap) to automatically determine track loudness (and other data like cue-in and cue-out points) _on-the-fly_. This produces extra CPU load (the values have to be calculated every time) but also ensures a smooth, same-loudness, and nicely transitioned playout. It can be used in a standalone LS script like so:
   - Make `cue_file` available in the path (requires Python3, ffmpeg, ffprobe).
   - Include `autocue.cue_file.liq` in your script, set preferences, activate it, and use `liq_amplify` instead of `replaygain_track_gain` for gain correction:
    ```
    %include "autocue.cue_file.liq"
    settings.autocue.cue_file.noclip := true  # clipping prevention
    settings.autocue.cue_file.nice := true  # Linux/MacOS only!
    settings.autocue.preferred := "cue_file"
    settings.autocue.amplify_behavior := "keep"
    enable_autocue_metadata()
    
    ...
    
    radio = amplify(1., override="liq_amplify", radio)
    ```
    Completing step (2) and/or (3) should already give perfect playout.

4. If you wish another playout loudness (not recommended, but often done in "loudness war" times), read the documentation and carefully adjust the settings.

5. If you _really_ want compression/limiting, you can add post-processing tools like [_master_me_](https://github.com/trummerschlunk/master_me) or [_Stereo Tool_](https://www.thimeo.com/stereo-tool/) to achieve a more fine-tuned sound. _Hint:_ Don’t overdo it!
   These tools are more easily integrated when using a "radio station in a box" like [_AzuraCast_](https://www.azuracast.com/) (which, in turn, also uses Liquidsoap internally).

6. [FLAC output](https://www.liquidsoap.info/doc-dev/encoding_formats.html#flac) is technically not a problem with Liquidsoap, but it has other issues (bandwidth consumption, in-stream metadata). If using that, I’d recommend using an Icecast stream server, and only using the FLAC stream inside the local LAN. You _can_ have multiple parallel streams from the same source using Liquidsoap (i.e., FLAC at home, MP3 or AAC for remote streams).

Have fun learning and experimenting!

**Reply by branch14** [reply]
Perfect write up! Thanks so much.
